### PR TITLE
[ready] collapse unnecessary instructions

### PR DESCRIPTION
**Pull request by MoKob**
This PR targets the collapse/removal of unnecessary instructions at segregated roads.

This includes name changes as well as the collapsing of two turns into a (possibly named) `u-turn`.

Contributing to https://github.com/Project-OSRM/osrm-backend/issues/2263

 - [ ] requires https://github.com/Project-OSRM/osrm-backend/pull/2248 merged
 - [x] review

### CODE REVIEW

File: src/engine/guidance/post_processing.cpp [hunk 1]
```diff
@@ -198,6 +198,66 @@ void print(const std::vector<RouteStep> &steps)
...
+
+        // elongating in the front changes the location of the maneuver
+        step.maneuver.location = by_step.maneuver.location;
+        step.maneuver.bearing_before = by_step.maneuver.bearing_before;
+        step.maneuver.bearing_after = by_step.maneuver.bearing_after;
+        step.maneuver.instruction = by_step.maneuver.instruction;
+
+        step.maneuver.intersections.insert(step.maneuver.intersections.begin(),
+                                           by_step.maneuver.intersections.begin(),
+                                           by_step.maneuver.intersections.end());
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: The insert call plus the return and is shared across branches with the `pos` iterator being different; maybe de-duplicate

File: src/engine/guidance/post_processing.cpp [hunk 2]
```diff
@@ -287,20 +347,152 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
...
+{
+    // invalidate a step and set its content to nothing
+    const auto invalidateStepAtIndex = [&steps](std::size_t index) {
+        auto &step = steps[index];
+        step.maneuver.instruction = TurnInstruction::NO_TURN();
+        step.duration = 0;
+        step.distance = 0;
+        step.geometry_begin = 0;
+        step.geometry_end = 0;
+        step.name.clear();
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: Hm is this the same as swapping step with an empty one? As in `step = RouteStep();`

File: src/engine/guidance/post_processing.cpp [hunk 3]
```diff
@@ -287,20 +347,152 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
...
+    // maneuvers that can, in some form, be seen as one. The additional in_step is to find out about
+    // a possible u-turn.
+    const auto collapsable = [](const RouteStep &step) {
+        const constexpr double MAX_COLLAPSE_DISTANCE = 25;
+        return step.distance < MAX_COLLAPSE_DISTANCE;
+    };
 
-    return steps;
+    const auto bearingsAreReversed = [](const double bearing_in, const double bearing_out) {
+        // Nearly perfectly refeversed angles have a difference close to 180 degrees (straight)
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: > refeversed

File: src/engine/guidance/post_processing.cpp (hunk 1 above)
Comment (MoKob) [inline src/engine/guidance/post_processing.cpp]: can do for the return. the insert would result in repeating the initial check, which would not reduce code complexity

File: src/engine/guidance/post_processing.cpp [hunk 4]
```diff
@@ -287,20 +347,152 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
...
+            {
+                // the simple case is a u-turn that changes directly into the in-name again
+                const bool direct_u_turn = steps[two_back_index].name == current_step.name;
+
+                // however, we might also deal with a dual-collapse scenario in which we have to
+                // additionall collapse a name-change as well
+                const bool u_turn_with_name_change =
+                    (step_index + 1 < steps.size()) && collapsable(current_step) &&
+                    (TurnType::NewName == steps[step_index + 1].maneuver.instruction.type) &&
+                    steps[step_index + 1].name == steps[two_back_index].name;
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: Maybe split this into named booleans.

File: src/engine/guidance/post_processing.cpp [hunk 5]
```diff
@@ -287,20 +347,152 @@ std::vector<RouteStep> postProcess(std::vector<RouteStep> steps)
...
+                    }
+
+                    steps[one_back_index].name = steps[two_back_index].name;
+                    steps[one_back_index].maneuver.instruction.type = TurnType::Continue;
+                    steps[one_back_index].maneuver.instruction.direction_modifier =
+                        DirectionModifier::UTurn;
+                }
+            }
+        }
+    }
```
Comment (daniel-j-h) [inline src/engine/guidance/post_processing.cpp]: The function is a bit of a monster, }}}}} looks like Lisp to me :P Maybe refactor a bit.

**Comment by TheMarex** [comment]
Please reopen the PR for `master`, `develop` is going away.

**Pull request closed without merging**